### ISSUE TITLE
create master resource with custom name

### ISSUE THREAD

**Issue by YanChii** [OndrejHome/ansible.pcs-modules-2#26]
Hi Ondrej,

I'd like to switch my pcs library to your modules in my [postresql cluster role](https://github.com/YanChii/ansible-role-postgres-ha). But I have one problem. Maybe you could help me with that.

I am creating a master/slave resource in two steps - [create resource](https://github.com/YanChii/ansible-role-postgres-ha/blob/master/tasks/paf.yml#L57) and [create master resource](https://github.com/YanChii/ansible-role-postgres-ha/blob/master/tasks/paf.yml#L94).
This is not possible with your `pcs_resource` module. Which is fine, I can do it in one task as you already [suggest in your example](https://github.com/OndrejHome/ansible.pcs-modules-2/blob/master/library/pcs_resource.py#L93).

But the problem is that it creates a resource with fixed name that appends `-master` after the original resource name (e.g. `postgres-master`). Is there any possiblity to have a custom master resource name? Because this is (unfortunately) what I have in my role and this role is already published... so changing the resource name would seriously mess already-installed user clusters.

I am also willing to create a PR to your library if necessary. But now I don't quite see any possiblity how to do that. Maybe there's a `pcs` option for that I am not aware of.

Any ideas?

Thank you for any thoughs.

Jan

**Comment by YanChii** [comment]
I've found a solution. I don't know if you like it but it works :).

https://github.com/OndrejHome/ansible.pcs-modules-2/compare/master...YanChii:custom-master-res-name

Jan

**Comment by OndrejHome** [comment]
Hello Jan,

Thank you for reaching out and opening the issue. Your assumption that there will be need to make some change to code is correct. I'm also not aware of any attribute that can be passed to `pcs resource create` with `master` specified that would make it to use the custom name of the Master/Slave resource.

Also thank you for the code proposal. I have not verified yet thoroughly, but will give it a try in upcoming days and let you know more about it. Having a look it seems good to fit the use case in `pcs-0.9`.
I have however concern for future use as currently the code for `pcs_resource` is aimed at `pcs-0.9` and there is Issue #20 for checking its compatibility with newer `pcs-0.10` version in Fedora 29 and RHEL 8. One major change that I see in contect of this issue is that there is no `pcs resource master` subcommand in `pcs-0.10` - it is replaced with `pcs resource promotable` and that command doesn't accept the alternative name for "Master/Slave"/Promotable resource - see the exert from usage doc below.

`pcs-0.9`
https://github.com/ClusterLabs/pcs/blob/09a539251213136736f839388b69a1cffa44c2ca/pcs/usage.py#L434
~~~
    master [<master/slave id>] <resource id | group id> [options] [--wait[=n]]
~~~

`pcs-0.10`
https://github.com/ClusterLabs/pcs/blob/a0912431ed23b10521dbd2a64d48dbce865be453/pcs/usage.py#L455
~~~
    promotable <resource id | group id> [clone options]... [--wait[=n]]
~~~

So in case of introducing this functionality in modules here it would not work for newer `pcs-0.10` version which I would like to get usable. 
I will need some time to check this problem further to see if there is any option in `pcs-0.10` to achieve same behaviour as with `pcs resource master my_name existing_resource` and will let you know. Feel free to check also on this if interested as this would affect you with future release `pcs-0.10` as you may face that naming issue with newer `pcs-0.10` - for completeness the `pcs resource promotable` creates resource name with `-clone` suffix to make it more "interesting" :)

I'm marking this Issue as enhancement and I will have a further look into this. At moment I'm  considering your contribution to be limited to `pcs-0.9` version or finding the solution that can also work with `pcs-0.10`. Once decided I will let you know so you can submit the PR. Thank you!

**Comment by OndrejHome** [comment]
OK, so from code of `pcs` it is pretty clear that while version 0.9 allowed custom names, the version 0.10 just hardcodes the `-clone` at the end and I don't see any direct `pcs` command for renaming this.

~~~
### pcs/resource.py pcs-0.9 @09a5392 Thu Feb 21 12:51:15 2019 +0100
1405 def resource_master_create(dom, argv, update=False, master_id=None):
1406     master_id_autogenerated = False
1407     if update:
1408         master_id = argv.pop(0)  <--- custom name possible
1409     elif not master_id:
1410         master_id = argv[0] + "-master"
1411         master_id_autogenerated = True
~~~

~~~
### pcs/resource.py master @a091243 Tue Mar 5 13:47:39 2019 +0100
1418 def resource_clone_create(
1419     cib_dom, argv, update_existing=False, promotable=False
1420 ):
...
1462     if update_existing:
1463         if element.parentNode.tagName != "clone":
1464             utils.err("%s is not currently a clone" % name)
1465         clone = element.parentNode
1466     else:
1467         clone = cib_dom.createElement("clone")
1468         clone.setAttribute("id", utils.find_unique_id(cib_dom, name + "-clone"))  <-- hard-coded name
1469         clone.appendChild(element)
1470         resources_el.appendChild(clone)
~~~

However I can still make change manually via CIB editing and push the ID of my liking there. So even if the "promotable" is created with `pcs resource promotable` or `pcs resource create ... promotable` I can use the `pcs cluster cib/cib-push` to get the editing of the final XML to match the name of my liking.
So in overall this will be possible to get there via `pcs` with note that `pcs-0.10` will use the editing of CIB XML which I try to avoid if possible.
I will start working on code for `pcs-0.10` with `promotable` function and once that is done I will let you know and you can prepare the PR for `pcs-0.9` version. 

For the final syntax I'm currently more inclined to idea of creating the master/slave resource with one invocation of `pcs_resource` command and not two, the syntax could be something like below:

~~~
- name: create Master resource 'test-ha' 
  pcs_resource:
    name: 'test-ha'
    resource_type: 'Stateful' (name of resource agent as seen with normal resource creation)
    child_name: 'test'  (this would be optional attribute)
    resource_class: 'master' (this will also accept word 'promotable' and maybe even 'clone')
    options: >
      some_attribute=value master meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true
~~~

@YanChii please let me know if the above approach with just single invocation of `pcs_resource` would match your use case. Looking at your role for postgresql cluster it would require merge of tasks 'create database cluster resource' and 'create master DB resource' into one tasks in `tasks/paf.yml`.

**Comment by YanChii** [comment]
Thank you Ondrej for your analysis and also for very interesting proposal. Single invocation of `pcs_resource` is fine for me, I'll make appropriate changes in my code.
Just drop me a message here and I'll take a look.
Thanks again.

Jan

PS: if the cib editing for some reason won't work, I can still fail my role if 
`pcs>= 0.10` and `ha_name !=~ /-master$/` 
and let the user to resolve.

**Comment by OndrejHome** [comment]
Hi Jan,

Thank you for confirmation. I will be looking at this over weekend and next week and will let you know once there is code ready for testing out.

Note that in `pcs-0.10` the master/slave resource is by default created with suffix `-clone` to denote that it is now "promotable clone" instead of `-master` used in `pcs-0.9` :)
The code also seems to accept automatic conversion of 'master/slave' resource from era of `pcs-0.9` into `pcs-0.10` era without changing the name or adding `-clone` - technically resource changes from 'master' to 'clone' but change is transparent to user/admin.

**Comment by OndrejHome** [comment]
Hi @YanChii ,
I've made some progress here with 2 issues at the moment that I believe should be solvable:

- [ ] Issue 1 - 'CIB too old' when renaming multistate resource
  - this can be workarounded with specifying option `force_resource_update: true` for that resource
  - I'm considering to make workaround the permanently enabled or find some other way
- [ ] Issue 2 - error when the multistate resource already exists, but something in it changed
  - this is issue with function that tries to show 'diff' that unfortunately hits this even when `--diff` is not specified
  - I will have a look at this over week, as that should be just issue with how to data are passed around

In the meantime feel free to check out the version from `develop` and see if that can be integrated in your project in form that is mentioned in the updated documentation of `pcs_resource` module. Creating and deleting multistate (Master/Slave) resources should work here. Updating them will fail (Issue 2). Please let me know if you have any additional feedback or issues to current version in the `develop` branch on the top of the 2 issues that I have described in this comment.

**Comment by YanChii** [comment]
Thank you, I'll take a look as soon as possible.
J.

**Comment by OndrejHome** [comment]
Hi @YanChii,
I have pushed to develop 2 changes to address the issues. The Issue 1 seems to appear only on my CentOS 7.6 (`pcs-0.9`) so I have added hint for using the `force_resource_update: true` into error message when that happens. On Fedora 29 cluster (`pcs-0.10`) I haven't got such issues.
Issue 2 seems to be localized to python 3.6 that I use on my system and I have changed one thing that should make it work also for me :) Anyway if you notice that something stopped working or started reporting suddenly error please let me know.

I will mark this Issue now to be working on your feedback if the current version is suitable for your deployment and also if you have encountered any issues. In case that I don't hear from you in next 2 weeks I would consider that there was no major issue and I will eventually merge this to `master` branch. In case that I have feedback from you it can happen faster or additional fixes can be added.

**Comment by YanChii** [comment]
Hi.

I've made the required changes in my role and everything seems to work perfectly. I didn't even have to add `force_resource_update: true` on `pcs` `0.9.165` @ `CentOS`. I'm gonna test if further before release.

Thank you very much for your help and effort. I really appreciate it.

Jan

**Comment by OndrejHome** [comment]
Hi @YanChii ,

Thank you for feedback. As I had also some other changes that I wanted to push to master I have merged this under tag `20.0.0`. In case that there are some issues they can be fixed and then published in another version. If not then this would be version that you can start using and refer to.

I will keep this issue open in case that you have any additional feedback to this. If not I will close it 2 weeks from now. Thank you once again for your contribution and feedback for this feature!

**Comment by OndrejHome** [comment]
Once again thank you @YanChii for your contribution here. As mentioned in my previous comment I'm closing this Issue now. In case you encounter any issue with current code in `master` please open new Issue and if it is related to changes discussed here please mention it.

Ondrej
